src/platform_impl/linux/x11/mod.rs

```diff
@@ -359,259 +403,304 @@ impl<T: 'static> EventLoop<T> {
+        })?;
 
-            // Empty the redraw requests
-            {
-                let mut windows = HashSet::new();
+        exit
+    }
 
-                // Empty the channel.
+    pub fn pump_events<F>(&mut self, timeout: Option<Duration>, mut callback: F) -> PumpStatus
+    where
+        F: FnMut(Event<'_, T>, &RootELW<T>, &mut ControlFlow),
+    {
+        if !self.loop_running {
+            self.loop_running = true;
 
-                while let Some(window_id) = this.state.redraw_events.pop_front() {
-                    windows.insert(window_id);
-                }
+            // Reset the internal state for the loop as we start running to
+            // ensure consistent behaviour in case the loop runs and exits more
+            // than once.
+            self.control_flow = ControlFlow::Poll;
 
-                for window_id in windows {
-                    let window_id = crate::window::WindowId(window_id);
-                    sticky_exit_callback(
-                        Event::RedrawRequested(window_id),
-                        &this.target,
-                        control_flow,
-                        callback,
-                    );
-                }
-            }
-            // send RedrawEventsCleared
-            {
-                sticky_exit_callback(
-                    crate::event::Event::RedrawEventsCleared,
-                    &this.target,
-                    control_flow,
-                    callback,
-                );
-            }
+            // run the initial loop iteration
+            self.single_iteration(&mut callback, StartCause::Init);
+        }
 
-            let start = Instant::now();
-            let (deadline, timeout);
+        // Consider the possibility that the `StartCause::Init` iteration could
+        // request to Exit.
+        if !matches!(self.control_flow, ControlFlow::ExitWithCode(_)) {
+            self.poll_events_with_timeout(timeout, &mut callback);
+        }
+        if let ControlFlow::ExitWithCode(code) = self.control_flow {
+            self.loop_running = false;
 
-            match control_flow {
-                ControlFlow::ExitWithCode(_) => {
-                    return IterationResult {
-                        wait_start: start,
-                        deadline: None,
-                        timeout: None,
-                    };
-                }
-                ControlFlow::Poll => {
-                    *cause = StartCause::Poll;
-                    deadline = None;
-                    timeout = Some(Duration::from_millis(0));
+            let mut dummy = self.control_flow;
+            sticky_exit_callback(
+                Event::LoopDestroyed,
+                self.window_target(),
+                &mut dummy,
+                &mut callback,
+            );
+
+            PumpStatus::Exit(code)
+        } else {
+            PumpStatus::Continue
+        }
+    }
```

Comment:
Why is the status of `self.control_flow == ControlFlow::ExitWithCode(_)` checked twice here?

Reply:
There was a bug here before that @kchibisov hit when testing that was due to the possibility that an application could _immediately_ ask to exit during the very first `Init` iteration of the loop that happens just above this quoted code.

So the first check is effectively accounting for `control_flow` being changed during the `Init` dispatch and it makes sure we don't poll if we already know the app wants to quit (which is important because it's currently considered a bug to reach polling when you know the app has asked to quit and will panic).

The second check is checking whether the `control_flow` was changed during dispatching within `poll_events_with_timeout`.